Add check for Content-Type when downloading tracker list

userwiths opened the pull request:
Closes #13061.

# Description

Some users might get confused around platforms like github and their `raw` links. By mistake such user has entered the HTML page link and each HTML line was taken as a tracker.
In order to fix this issue, we add a check on the `Content-Type` header of the response. We accept `text/plain` and show a descriptive error message otherwise.

userwiths committed:
fix: Add check for Content-Type when downloading tracker list.

Check if the URL from where we are downloading the trackers returns `Content-Type` as `text`.
Otherwise consider it might be a non-usable format for this functionality (html,xml,etc.ect.)

Chocobo1 commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -109,6 +109,13 @@ void TrackersAdditionDialog::onTorrentListDownloadFinished(const Net::DownloadRe
     m_ui->downloadButton->setEnabled(true);
     setCursor(Qt::ArrowCursor);
 
+    if (!result.contentType.contains(u"text/plain"_s, Qt::CaseInsensitive))
```
Shouldn't you move it after `if (result.status != Net::DownloadStatus::Success) {}`?

userwiths replied:
Correct, oversight on my part and addressed in the latest commit.

glassez commented:
> Some users might get confused around platforms like github and their `raw` links. By mistake such user has entered the HTML page link and each HTML line was taken as a tracker.
> In order to fix this issue, we add a check on the `Content-Type` header of the response. We accept `text/plain` and show a descriptive error message otherwise.

The plain text may also contain arbitrary/inappropriate data. Shouldn't the lines themselves be validated instead?

userwiths committed:
Check header after request succeeds, check if added trackers are valid URL\'s

userwiths commented:
I thought that the text area was meant to be used in order to "preview" the url's a user was about to add.
But I agree that we could limit their chances of entering gibberish accidentally.
Latest commit adds a simple check to see if the text looks like a url.

stalkerok commented:
Doesn't work, incorrect list added without a problem.

userwiths committed:
Add check to Bittorrent tab in preferences too.

stalkerok commented:
Tested!

userwiths commented:
@stalkerok @glassez 
I see that there is a stale-bot, I'm wondering if there is anything I could do in order to close this PR "naturally" (Merged or Closed due to issues) or keep it open until then, rather than have the bot close it automatically.

luzpaz commented:
What's left to do here ?

userwiths commented:
A review maybe.
Don't know if this comment (https://github.com/qbittorrent/qBittorrent/pull/22137#issuecomment-2598672788) counts as such or not.

glassez commented on src/gui/optionsdialog.h:
```diff
@@ -34,6 +34,7 @@
 #include "base/pathfwd.h"
 #include "base/settingvalue.h"
 #include "guiapplicationcomponent.h"
+#include "base/net/downloadmanager.h"
```
Wrong order.

glassez commented on src/gui/optionsdialog.cpp:
```diff
@@ -83,6 +83,7 @@
 #include "watchedfolderoptionsdialog.h"
 #include "watchedfoldersmodel.h"
 #include "webui/webui.h"
+#include "base/net/downloadmanager.h"
```
It is even redundant since it is included in `optionsdialog.h`.

glassez commented on src/gui/optionsdialog.h:
```diff
@@ -137,9 +138,12 @@ private slots:
     void loadSearchTabOptions();
     void saveSearchTabOptions() const;
 
+
```
```suggestion
```

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,14 +74,35 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
     delete m_ui;
 }
 
+int TrackersAdditionDialog::isValidEndpoint(const QStringView &endpoint) const
```
```suggestion
bool TrackersAdditionDialog::isValidEndpoint(const QStringView endpoint) const
```

...and return `false` or `true`.

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,14 +74,35 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
+{
+    if (endpoint.isEmpty())
+        return 0;
+    QUrl url(endpoint.toString());
```
```suggestion
    const QUrl url {endpoint.toString()};
```

userwiths replied:
Changed it, should be ok now.

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,14 +74,35 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
 
     QList<BitTorrent::TrackerEntry> entries = BitTorrent::parseTrackerEntries(m_ui->textEditTrackersList->toPlainText());
-    for (BitTorrent::TrackerEntry &entry : entries)
+    for (BitTorrent::TrackerEntry &entry : entries) {
```
Broken coding style.

```suggestion
    for (BitTorrent::TrackerEntry &entry : entries)
    {
```

userwiths replied:
Fixed

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,14 +74,35 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
     QList<BitTorrent::TrackerEntry> entries = BitTorrent::parseTrackerEntries(m_ui->textEditTrackersList->toPlainText());
-    for (BitTorrent::TrackerEntry &entry : entries)
+    for (BitTorrent::TrackerEntry &entry : entries) {
+        auto isValid = isValidEndpoint(entry.url);
```
Seems this variable doesn't make much sense.
```suggestion
       if (isValidEndpoint(entry.url))
```

userwiths replied:
Yeah, you are right, addressed.

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,14 +74,35 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
+        auto isValid = isValidEndpoint(entry.url);
+        if (!isValid)
+        {
+            QMessageBox::warning(const_cast<TrackersAdditionDialog *>(this), tr("Invalid tracker URL"), tr("The tracker URL \"%1\" is invalid").arg(entry.url));
```
I would drop `const` specifier from `onAccepted()` than use `const_cast` here.

userwiths replied:
Sure thing, done.

glassez commented on src/gui/trackersadditiondialog.h:
```diff
@@ -65,6 +65,7 @@ private slots:
 private:
     void saveSettings();
     void loadSettings();
+    int isValidEndpoint(const QStringView &endpoint) const;
```
This function is independent of `TrackersAdditionDialog` class, so it can be declared in an anonymous namespace in the `trackersadditiondialog.cpp` file.

userwiths replied:
Fair enough, made it an anonymous function.

glassez reviewed (commented):
Perhaps I haven't identified all the problems yet...

userwiths committed:
Address PR comments.

userwiths committed:
Remove redundant include.

userwiths commented on src/gui/optionsdialog.cpp:
```diff
@@ -1229,8 +1230,42 @@ void OptionsDialog::saveBittorrentTabOptions() const
     session->setAddTrackersFromURLEnabled(m_ui->checkAddTrackersFromURL->isChecked());
     session->setAdditionalTrackersURL(m_ui->textTrackersURL->text());
+
+    return;
```
Right, wrote an explicit return by habit, will make sure it doesn't happen in the future.

userwiths commented:
Ok, all points should be covered now, even those on which I did not comment.

In case there is anything else, feel free to send it my way, im gonna take a look as soon as possible.

PS: thx for the time.

glassez commented:
Well, here's my general review.

I have no particular complaints about the `TrackersAdditionDialog`.

As for the main option, its processing is done in a dubious way, IMO.
To begin with, it would be more correct to check downloaded trackers list directly in the place where it is processed by the BitTorrent Session, and reject unsuitable data with a message entry in the log.
A preliminary check in the Options dialog may look like a good idea, but it is not implemented correctly. IMO, it is a bad idea to initiate data downloading when applying options. Wouldn't it be better to add a "Test" button to do this explicitly? Otherwise, we have two cases:
1. The user clicked the "Apply" button.  In this case, there is a chance that user will see a corresponding message about an unsuccessful download of trackers.
3. The user clicked the "Ok" button which closes the dialog immediatelly after applying the options. This is at least a race condition. Most likely, the user will not receive any message at all, since the dialog will be destroyed before that.

userwiths committed:
fix: As advised, move logic away from the settings window into session.

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,14 +74,35 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
 }
 
-void TrackersAdditionDialog::onAccepted() const
+bool isValidEndpoint(const QStringView endpoint)
```
This function should be placed inside anonymous namespace above `TrackersAdditionDialog` definitions.

userwiths replied:
Probably misunderstood you in my last commit but ....
Ive move the `isValidEndpoint` into a anonym. namespace but in the `.cpp` file.
`definitions` (from your comment) would imply the `.h`. Im ready to move it again in case its required.

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,14 +74,35 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
+        if (!isValidEndpoint(entry.url))
+        {
+            QMessageBox::warning(this, tr("Invalid tracker URL"), tr("The tracker URL \"%1\" is invalid").arg(entry.url));
+            return;
```
IMO, it looks unfriendly if you reject all the entered trackers because of invalid one. Or am I misunderstood it?

userwiths replied:
Now that i think about it, that check runs for the `txt` downloaded trackers too.
Yeah, my bad, changed the return so that the info is still logged but `entry`'s do not get skipped cause of the first bad one.

glassez commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -4105,6 +4105,11 @@ void SessionImpl::updateTrackersFromURL()
             {
+                if (!result.contentType.contains(u"text/plain"_s, Qt::CaseInsensitive))
+                {
+                    LogMsg(tr("Cannot add trackers from URL, expected Content-Type is \'text/plain\' received \"%1\"").arg(result.contentType), Log::WARNING);
```
1. Single and double quotes are used inconsistently in this message.
2. I doubt that single quotes need to be escaped.
3. The message format seems inappropriate (and even grammatically invalid).

@Chocobo1?

userwiths replied:
1. Fixed, now we use double quotes just as the other Log msg's
2. Ur probably right, over-caution on my side
3. Only thing that can be changed imo is the `is` => `was`, im open to suggestions, but i have no such to provide atm

stalkerok commented:
Related of

- https://github.com/qbittorrent/qBittorrent/pull/23939
- https://github.com/arvidn/libtorrent/pull/8202

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -46,6 +46,22 @@
 
 #define SETTINGS_KEY(name) u"AddTrackersDialog/" name
 
+namespace {
```
Invalid coding style!
```suggestion
namespace
{
```

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,14 +90,21 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
+    {
+        if (!isValidEndpoint(entry.url))
+        {
+            QMessageBox::warning(this, tr("Invalid tracker URL"), tr("The tracker URL \"%1\" is invalid").arg(entry.url));
```
```suggestion
            QMessageBox::warning(this, tr("Invalid tracker URL"), tr("The tracker URL \"%1\" is invalid.").arg(entry.url));
```

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,14 +90,21 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
+    {
+        if (!isValidEndpoint(entry.url))
+        {
+            QMessageBox::warning(this, tr("Invalid tracker URL"), tr("The tracker URL \"%1\" is invalid").arg(entry.url));
```
What if a lot of URLs are invalid? The user will be swamped with a lot of message boxes. Wouldn't it be better to accumulate the overall result, for example, to show the first error message and the number of additional errors?

userwiths replied:
<img width="441" height="181" alt="image" src="https://github.com/user-attachments/assets/90f9d6e1-3fd8-429d-bf1d-5c0dda235327" />

:+1:

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -116,6 +139,13 @@ void TrackersAdditionDialog::onTorrentListDownloadFinished(const Net::DownloadRe
+    if (!result.contentType.contains(u"text/plain"_s, Qt::CaseInsensitive))
+    {
+        QMessageBox::warning(this, tr("Download trackers list error")
+            , tr("The content type of the downloaded file is not plain text. Content-Type: \"%1\"").arg(result.contentType));
```
```suggestion
            , tr("The content type of the downloaded file is not \"text/plain\". Content-Type: \"%1\".").arg(result.contentType));
```

glassez commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -4072,6 +4072,11 @@ void SessionImpl::updateTrackersFromURL()
             {
+                if (!result.contentType.contains(u"text/plain"_s, Qt::CaseInsensitive))
+                {
+                    LogMsg(tr("Cannot add trackers from URL, expected Content-Type was \"text/plain\" received \"%1\"").arg(result.contentType), Log::WARNING);
```
I still don't like this message. Wouldn't something like the following be better?
`Cannot add trackers from URL due to unexpected Content-Type. Expected: \"text/plain\". Received: \"%1\".`

userwiths replied:
Both variations seem informative enough to me, im gonna follow your lead and change it to the suggested message.

glassez commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -4072,6 +4072,11 @@ void SessionImpl::updateTrackersFromURL()
             {
+                if (!result.contentType.contains(u"text/plain"_s, Qt::CaseInsensitive))
+                {
+                    LogMsg(tr("Cannot add trackers from URL due to unexpected Content-Type. Expected: \"text/plain\". Received: \"%1\".").arg(result.contentType), Log::WARNING);
```
```suggestion
                    LogMsg(tr("Cannot add trackers from URL due to unexpected Content-Type. Expected: \"text/plain\". Received: \"%1\".")
                            .arg(result.contentType), Log::WARNING);
```

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,15 +91,28 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
+    }
 
+    if (invalidTrackersCount > 0)
+        QMessageBox::warning(this, tr("Invalid tracker URL"), tr("The tracker URL \"%1\" and %2 other(s) are invalid.").arg(firstInvalidTracker).arg(invalidTrackersCount - 1));
```
```suggestion
    if (invalidTrackersCount > 1)
    {
        QMessageBox::warning(this, tr("Invalid tracker URL")
            , tr("The tracker URL \"%1\" and %2 other(s) are invalid.").arg(firstInvalidTracker).arg(invalidTrackersCount - 1));
    }
    else if (invalidTrackersCount == 1)
    {
        QMessageBox::warning(this, tr("Invalid tracker URL")
            , tr("The tracker URL \"%1\" is invalid.").arg(firstInvalidTracker));
    }
```

userwiths replied:
Ok we now have a message for each case (>1 and =1) and the formatting in both places should be fixed.

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,15 +91,36 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
+    if (invalidTrackersCount > 1)
+    {
+        QMessageBox::warning(this, tr("Invalid tracker URL")
+            , tr("The tracker URL \"%1\" and %2 other(s) are invalid.").arg(firstInvalidTracker).arg(invalidTrackersCount - 1));
```
For better translation it should be refactored as the following:
```suggestion
            , tr("The tracker URL \"%1\" and %n other(s) are invalid.", "The tracker URL \"foobar\" and 2 others are invalid.", (invalidTrackersCount - 1)).arg(firstInvalidTracker));
```

userwiths replied:
Well , i tested this locally and im not able to translate it if i use the suggested format/code.

The current text im able to translate with
```
<message>
        <location filename="../gui/trackersadditiondialog.cpp" line="117"/>
        <source>The tracker URL &quot;%1&quot; and %2 other(s) are invalid.</source>
        <translation>Тракер с адрес &quot;%1&quot; и %2 други са невалидни.</translation>
</message>
```

And im trying to translate the suggested text with 

```
<message>
        <location filename="../gui/trackersadditiondialog.cpp" line="117"/>
        <source>The tracker URL &quot;%1&quot; and %n other(s) are invalid.</source>
        <translation>Тракер с адрес &quot;%1&quot; и %n други са невалидни.</translation>
</message>
```

also tried using the second string-arg for translation, still not translating:
```
<message>
        <location filename="../gui/trackersadditiondialog.cpp" line="117"/>
        <source>The tracker URL &quot;foobar&quot; and 2 others are invalid.</source>
        <translation>Тракер с адрес &quot;%1&quot; и %n други са невалидни.</translation>
</message>
```

I dont see the `%n` trick used anywhere else in the translation files (i see `%N` but its not a translate, its a literal)

Maybe im missing something , but would refrain from changing it without being able to translate it.

glassez replied:
> Maybe im missing something , but would refrain from changing it without being able to translate it.

Did you update translation sources using `lupdate`?

Chocobo1 commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -4072,6 +4072,12 @@ void SessionImpl::updateTrackersFromURL()
         {
             if (result.status == Net::DownloadStatus::Success)
             {
+                if (!result.contentType.contains(u"text/plain"_s, Qt::CaseInsensitive))
```
```suggestion
                if (!result.contentType.contains(u"text/plain", Qt::CaseInsensitive))
```

userwiths replied:
Thats the only thing im not 100% sure why the `_s` is removed. I guess its something to do with how much the the text is used in the code. But the change is applied either way.

Chocobo1 commented on src/gui/optionsdialog.cpp:
```diff
@@ -50,6 +50,7 @@
 #include "base/bittorrent/sharelimits.h"
 #include "base/exceptions.h"
 #include "base/global.h"
+#include "base/net/downloadmanager.h"
```
Seems unused:
```suggestion
```

Chocobo1 commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -46,6 +46,23 @@
+    {
+        if (endpoint.isEmpty())
+            return false;
+        const QUrl url {endpoint.toString()};
```
The fact you called `QStringView::toString()` means you should use `QString &` to avoid a deep copy:
```suggestion
    bool isValidEndpoint(const QString &endpoint)
    {
        if (endpoint.isEmpty())
            return false;
        const QUrl url {endpoint};
```

Chocobo1 commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -116,6 +154,13 @@ void TrackersAdditionDialog::onTorrentListDownloadFinished(const Net::DownloadRe
+    if (!result.contentType.contains(u"text/plain"_s, Qt::CaseInsensitive))
+    {
+        QMessageBox::warning(this, tr("Download trackers list error")
+            , tr("The content type of the downloaded file is not \"text/plain\". Content-Type: \"%1\".").arg(result.contentType));
```
```suggestion
            , tr("The content type of the downloaded file is not \"text/plain\". Received Content-Type: \"%1\".").arg(result.contentType));
```

userwiths commented:
Latest push contains the translation related suggestion from https://github.com/qbittorrent/qBittorrent/pull/22137#discussion_r3289591043

glassez reviewed (dismissed):
@qbittorrent/bug-handlers 
I hope that one of you (each of you) tested this one last time. Unfortunately, I don't have time for this (even sometimes for detailed testing of my own PRs).

stalkerok commented:
It looks like there are some issues.

An error appears in the log, but the unexpected content type is still added when the torrent is added.

userwiths commented:
I think i know how/when this happens, but i have no idea for a work around.

In "Preferences" the field "Fetched tracker" (see img bellow) is "saved". That is after a first initial load of the url the download is **not** repeated in next session. So if u have managed to get it populated with bad data (like, if you populated it with from an `html` before checking-out this branch) the download does not trigger, no check because no download, and the already saved "Fetched trackers" get added to the new torrent.

In short, people who have added a bad link before the fix, will keep seeing bad trackers after the fix because the trackers are saved and not re-downloaded again.

<img width="618" height="215" alt="image" src="https://github.com/user-attachments/assets/43491741-33c9-4835-af15-47bfe968730c" />

stalkerok commented:
https://github.com/user-attachments/assets/5d684b59-1d49-46f5-a24f-557d5fd42c2d

userwiths committed:
fix: if content-type is invalid, delete trackers file that was saved

userwiths commented:
Ok, so i've missed the part where the trackers (from URL) get downloaded and loaded from the local file on the next session-start.
Last commit checks the result from the `Content-Type` check and deletes the local file in case its not `text/plain`.

Chocobo1 commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -4067,10 +4067,17 @@ void SessionImpl::updateTrackersFromURL()
         {
             if (result.status == Net::DownloadStatus::Success)
             {
+                if (!result.contentType.contains(u"text/plain", Qt::CaseInsensitive))
```
Not sure if someone already asked about it.
Why check the content type from the server instead of checking the data itself?
For example, you can use `QMimeDatabase::mimeTypeForData()` for this purpose.

glassez replied:
> Not sure if someone already asked about it.
> Why check the content type from the server instead of checking the data itself?

Checking header value is cheaper, no? And it is enough in east majority of the cases (except the malicious server ones).

Chocobo1 replied:
>And it is enough in east majority of the cases 

Are you sure? There are a lot of problematic servers in the wild from my experience.

IMO the only way to ensure data correctness is to parse the received data properly: check whether each line is a valid tracker URL. And reject the data on any error.

glassez replied:
> Are you sure? There are a lot of problematic servers in the wild from my experience.

Well, then it should be applicable to every HTTP response received. But no one does that. In the end, the data is still processed by the appropriate parser, so that's its concern.

> IMO the only way to ensure data correctness is to parse the received data properly: check whether each line is a valid tracker URL. And reject the data on any error.

👍

Chocobo1 replied:
> > IMO the only way to ensure data correctness is to parse the received data properly: check whether each line is a valid tracker URL. And reject the data on any error.
> 
> 👍

Still, I would prefer to not have this PR since it doesn't really reject invalid tracker list and only checking Content-type header is still insufficient and error prone.
If others think this PR is good then merge it as you see fit. I won't participate.

glassez replied:
I believe the original intention was to prevent inappropriate formatted data from the respectable sites (e.g. if the user has specified a website address that returns an HTML page containing a list of trackers, so this is not suitable for use).

@qbittorrent/bug-handlers 
Should we reject this PR and insist that a strict parser of the tracker list be implemented in order for it to be approved? Or can we merge it as is and let someone else improve it later?

stalkerok replied:
The current implementation will cover most cases of incorrect data being added. I think that's acceptable.

glassez commented on src/gui/trackersadditiondialog.cpp:
```diff
@@ -74,15 +91,36 @@ TrackersAdditionDialog::~TrackersAdditionDialog()
+        QMessageBox::warning(this, tr("Invalid tracker URL")
+            , tr("The tracker URL \"%1\" is invalid.").arg(firstInvalidTracker));
+    }
     m_torrent->addTrackers(entries);
```
So, are you adding them anyway, even if they contain invalid ones?

userwiths replied:
Not the intention, `continue;`s got removed [here](https://github.com/qbittorrent/qBittorrent/pull/22137#discussion_r3259607429) most likely, will fix when i get a chance.

userwiths replied:
Should be handled now, first tried with `std:views:filter` but it seems some of the CI's did not like it
So we just create an empty `QList` where we put the **valid** entries.

userwiths committed:
fix: Add only the trackers that pass the url check

> first attempt was with `std::views` but it seems older C++ CI does not like it
> using simpler approach (new list and append valid entries)
